File: pkl-formatter/src/main/java/org/pkl/formatter/Generator.java
/*
 * Copyright © 2026 Apple Inc. and the Pkl project authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.pkl.formatter;

import java.util.HashSet;
import java.util.Set;

final class Generator {

  static final int MAX = 100;
  private static final String INDENT = "  ";

  private final Appendable buf;
  private int indent = 0;
  private int size = 0;
  private final Set<Integer> wrapped = new HashSet<>();
  private boolean shouldAddIndent = false;

  Generator(Appendable buf) {
    this.buf = buf;
  }

  void generate(FormatNode node) {
    node(node, Wrap.DETECT);
  }

  @SuppressWarnings("StatementWithEmptyBody")
  private void node(FormatNode node, Wrap wrap) {
    if (node instanceof Empty) {
      // nothing
    } else if (node instanceof Nodes n) {
      for (var child : n.nodes()) {
        node(child, wrap);
      }
    } else if (node instanceof Group g) {
      var width = 0;
      for (var child : g.nodes()) {
        width += child.width(wrapped);
      }
      var groupWrap = wrap;
      if (size + width > MAX) {
        wrapped.add(g.id());
        groupWrap = Wrap.ENABLED;
      } else {
        groupWrap = Wrap.DETECT;
      }
      for (var child : g.nodes()) {
        node(child, groupWrap);
      }
    } else if (node instanceof IfWrap iw) {
      if (wrapped.contains(iw.id())) {
        node(iw.ifWrap(), Wrap.ENABLED);
      } else {
        node(iw.ifNotWrap(), wrap);
      }
    } else if (node instanceof Text t) {
      text(t.text());
    } else if (node instanceof Line) {
      if (wrap.isEnabled()) {
        newline(true);
      }
    } else if (node instanceof ForceLine) {
      newline(true);
    } else if (node instanceof SpaceOrLine) {
      if (wrap.isEnabled()) {
        newline(true);
      } else {
        text(" ");
      }
    } else if (node instanceof Space) {
      text(" ");
    } else if (node instanceof Indent ind) {
      if (wrap.isEnabled() && !ind.nodes().isEmpty()) {
        size += INDENT.length();
        indent++;
        for (var child : ind.nodes()) {
          node(child, wrap);
        }
        indent--;
      } else {
        for (var child : ind.nodes()) {
          node(child, wrap);
        }
      }
    } else if (node instanceof MultilineStringGroup multi) {
      var indentLength = indent * INDENT.length();
      var oldIndent = indentFor(multi);
      var previousNewline = false;
      var nodes = multi.nodes();
      for (var i = 0; i < nodes.size(); i++) {
        var child = nodes.get(i);
        if (child instanceof ForceLine) {
          newline(false);
        } else if (child instanceof Text t
            && previousNewline
            && t.text().isBlank()
            && t.text().length() == oldIndent
            && nodes.get(i + 1) instanceof ForceLine) {
          // skip blank line indentation that will be repositioned
        } else if (child instanceof Text t && previousNewline) {
          text(reposition(t.text(), multi.endQuoteCol() - 1, indentLength));
        } else {
          node(child, Wrap.DETECT);
        }
        previousNewline = child instanceof ForceLine;
      }
    }
  }

  private void text(String value) {
    try {
      if (shouldAddIndent) {
        for (var i = 0; i < indent; i++) {
          buf.append(INDENT);
        }
        shouldAddIndent = false;
      }
      size += value.length();
      buf.append(value);
    } catch (java.io.IOException e) {
      throw new java.io.UncheckedIOException(e);
    }
  }

  private void newline(boolean shouldIndent) {
    try {
      size = INDENT.length() * indent;
      buf.append('\n');
      shouldAddIndent = shouldIndent;
    } catch (java.io.IOException e) {
      throw new java.io.UncheckedIOException(e);
    }
  }

  // accept text indented by originalOffset characters (tabs or spaces)
  // and return it indented by newOffset characters (spaces only)
  private static String reposition(String text, int originalOffset, int newOffset) {
    return " ".repeat(newOffset) + text.substring(originalOffset);
  }

  private static int indentFor(MultilineStringGroup multi) {
    var nodes = multi.nodes();
    if (nodes.size() < 2) return 0;
    var beforeLast = nodes.get(nodes.size() - 2);
    return beforeLast instanceof Text t ? t.text().length() : 0;
  }
}

File: pkl-formatter/src/main/kotlin/org/pkl/formatter/Generator.kt
/*
 * Copyright © 2025 Apple Inc. and the Pkl project authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.pkl.formatter

import org.pkl.formatter.ast.Empty
import org.pkl.formatter.ast.ForceLine
import org.pkl.formatter.ast.FormatNode
import org.pkl.formatter.ast.Group
import org.pkl.formatter.ast.IfWrap
import org.pkl.formatter.ast.Indent
import org.pkl.formatter.ast.Line
import org.pkl.formatter.ast.MultilineStringGroup
import org.pkl.formatter.ast.Nodes
import org.pkl.formatter.ast.Space
import org.pkl.formatter.ast.SpaceOrLine
import org.pkl.formatter.ast.Text
import org.pkl.formatter.ast.Wrap

internal class Generator {
  private val buf: StringBuilder = StringBuilder()
  private var indent: Int = 0
  private var size: Int = 0
  private val wrapped: MutableSet<Int> = mutableSetOf()
  private var shouldAddIndent = false

  fun generate(node: FormatNode) {
    node(node, Wrap.DETECT)
  }

  private fun node(node: FormatNode, wrap: Wrap) {
    when (node) {
      is Empty -> {}
      is Nodes -> node.nodes.forEach { node(it, wrap) }
      is Group -> {
        val width = node.nodes.sumOf { it.width(wrapped) }
        val wrap =
          if (size + width > MAX) {
            wrapped += node.id
            Wrap.ENABLED
          } else {
            Wrap.DETECT
          }
        node.nodes.forEach { node(it, wrap) }
      }
      is IfWrap -> {
        if (wrapped.contains(node.id)) {
          node(node.ifWrap, Wrap.ENABLED)
        } else {
          node(node.ifNotWrap, wrap)
        }
      }
      is Text -> text(node.text)
      is Line -> {
        if (wrap.isEnabled()) {
          newline()
        }
      }
      is ForceLine -> newline()
      is SpaceOrLine -> {
        if (wrap.isEnabled()) {
          newline()
        } else {
          text(" ")
        }
      }
      is Space -> text(" ")
      is Indent -> {
        if (wrap.isEnabled() && node.nodes.isNotEmpty()) {
          size += INDENT.length
          indent++
          node.nodes.forEach { node(it, wrap) }
          indent--
        } else {
          node.nodes.forEach { node(it, wrap) }
        }
      }
      is MultilineStringGroup -> {
        val indentLength = indent * INDENT.length
        val oldIndent = indentFor(node)
        var previousNewline = false
        for ((i, child) in node.nodes.withIndex()) {
          when {
            child is ForceLine -> newline(shouldIndent = false) // don't indent
            child is Text &&
              previousNewline &&
              child.text.isBlank() &&
              child.text.length == oldIndent &&
              node.nodes[i + 1] is ForceLine -> {}
            child is Text && previousNewline ->
              text(reposition(child.text, node.endQuoteCol - 1, indentLength))
            else -> node(child, Wrap.DETECT) // always detect wrapping
          }
          previousNewline = child is ForceLine
        }
      }
    }
  }

  private fun text(value: String) {
    if (shouldAddIndent) {
      repeat(times = indent) { buf.append(INDENT) }
      shouldAddIndent = false
    }
    size += value.length
    buf.append(value)
  }

  private fun newline(shouldIndent: Boolean = true) {
    size = INDENT.length * indent
    buf.append('\n')
    shouldAddIndent = shouldIndent
  }

  // accept text indented by originalOffset characters (tabs or spaces)
  // and return it indented by newOffset characters (spaces only)
  private fun reposition(text: String, originalOffset: Int, newOffset: Int): String =
    " ".repeat(newOffset) + text.drop(originalOffset)

  // Returns the indent of this multiline string, which is the size of the last node before the
  // closing quotes, or 0 if the closing quotes have no indentation
  private fun indentFor(multi: MultilineStringGroup): Int {
    val nodes = multi.nodes
    if (nodes.size < 2) return 0
    val beforeLast = nodes[nodes.lastIndex - 1]
    return if (beforeLast is Text) beforeLast.text.length else 0
  }

  override fun toString(): String {
    return buf.toString()
  }

  companion object {
    // max line length
    const val MAX = 100
    private const val INDENT = "  "
  }
}
